applications/nrf_desktop/src/modules/led_stream.c

```diff
@@ -0,0 +1,241 @@
+		     % CONFIG_DESKTOP_LED_STREAM_QUEUE_SIZE;
+	return CONFIG_DESKTOP_LED_STREAM_QUEUE_SIZE - len;
+}
+
+static bool store_data(const struct event_dyndata *data)
+{
+	size_t free_places = count_free_places();
+
+	if (free_places > 0) {
+		LOG_INF("Insert data on position %" PRIu8
+			", free places %zu", rx_idx, free_places);
+
+		if (!queue_data(data)) {
+			return false;
+		}
+	} else {
+		LOG_WRN("Queue is full - dropping incoming step");
+	}
+	return true;
+}
+
+static void send_data_from_queue(void)
+{
+	size_t free_places = count_free_places();
+
+	if (free_places < CONFIG_DESKTOP_LED_STREAM_QUEUE_SIZE) {
+		LOG_INF("Sending data");
+
+		if (first) {
+			first_led_stream_effect.steps = &steps_queue[tx_idx];
+			send_effect(&first_led_stream_effect, STREAM_LED_ID);
+			first = false;
+		} else {
+			second_led_stream_effect.steps = &steps_queue[tx_idx];
+			send_effect(&second_led_stream_effect, STREAM_LED_ID);
+			first = true;
+		}
+		tx_idx = next_index(tx_idx);
+	} else {
+		LOG_INF("No steps ready in queue");
+		first_sent = false;
+	}
+}
+
+static void handle_incoming_step(const struct config_event *event)
+{
+	if (!initialized) {
+		LOG_WRN("Not initialized");
+		return;
+	}
+
+	if (!store_data(&event->dyndata)) {
+		return;
+	}
+
+	if (!first_sent) {
+		LOG_INF("Sending first led effect");
+
+		send_data_from_queue();
+		first_sent = true;
+	}
+}
+
+static void handle_config_event(const struct config_event *event)
+{
+	if (!event->store_needed) {
+		/* Accept only events coming from transport. */
+		return;
+	}
+
+	if (GROUP_FIELD_GET(event->id) != EVENT_GROUP_LED_STREAM) {
+		/* Only LED STREAM events. */
+		return;
+	}
+
+	switch (TYPE_FIELD_GET(event->id)) {
+	case STREAM_DATA:
+		handle_incoming_step(event);
+		break;
+
+	default:
+		LOG_WRN("Unknown LED STREAM config event");
+		break;
+	}
+}
+
+static void handle_config_fetch_request_event(
+	const struct config_fetch_request_event *event)
+{
+	LOG_INF("Handle config fetch request event");
+
+	if (GROUP_FIELD_GET(event->id) != EVENT_GROUP_LED_STREAM) {
+		/* Only LED STREAM events. */
+		return;
+	}
+
+	u8_t free_places = count_free_places();
+
+	LOG_INF("Free places left: %" PRIu8, free_places);
+
+	struct config_fetch_event *fetch_event =
+		new_config_fetch_event(sizeof(free_places));
+
+	fetch_event->id = event->id;
+	fetch_event->recipient = event->recipient;
+	fetch_event->channel_id = event->channel_id;
+
+	fetch_event->dyndata.data[0] = free_places;
+
+	EVENT_SUBMIT(fetch_event);
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE)) {
+		if (is_config_event(eh)) {
+			handle_config_event(cast_config_event(eh));
+			return false;
+		}
+		if (is_config_fetch_request_event(eh)) {
+			handle_config_fetch_request_event(
+				cast_config_fetch_request_event(eh));
+			return false;
+		}
+	}
+
+	if (is_led_ready_event(eh)) {
+		send_data_from_queue();
```

Comment:
there is no check which led is ready...
there is no infrastructure to handle streams to different leds...

Reply:
done

Reply:
done